publicgen/reflect.go

```diff
@@ -0,0 +1,80 @@
+package publicgen
+
+import (
+	"fmt"
+	"reflect"
+	"strings"
+	"unicode"
+)
+
+var typeStruct = map[string]string{
+	"bson.ObjectId": "string",
+}
+
+func listFields(st reflect.Type) ([]field, []reflect.Type) {
+	fields := make([]field, 0)
+	imported := make([]reflect.Type, 0)
+
+	for i := 0; i < st.NumField(); i++ {
+		curField := st.Field(i)
+
+		if curField.Anonymous {
+			newFields, newTypes := listFields(st.Field(i).Type)
+			fields = append(fields, newFields...)
+			imported = append(imported, newTypes...)
+			continue
+		}
+
+		name := curField.Name
+		if unicode.IsLower([]rune(name)[0]) {
+			continue
+		}
+
+		json := curField.Tag.Get("json")
+		slice := false
+		typ := curField.Type
+
+		if curField.Type.Kind() == reflect.Slice {
+			slice = true
+			typ = curField.Type.Elem()
+		}
+
+		pointer := false
+		if typ.Kind() == reflect.Ptr {
+			typ = typ.Elem()
+			pointer = true
+		}
+
+		pkg := typ.PkgPath()
+		typePrefix := ""
+
+		if pkg != "" {
+			path := strings.Split(pkg, "/")
+			typePrefix = path[len(path)-1] + "."
```

Comment:
Would be nice to have a comment with an example to better understand this string manipulation